images/cfw/opt/x1plus/lib/python/x1plus/utils.py

```diff
@@ -37,3 +54,106 @@ def get_IP() -> str:
+        self.loaded = False
+        self.path = path
+        self.package_name = package
+        self.driver_data = module_docstring_parser(path, loader_type)
+        if not self.driver_data:
+            raise ImportError("module has no docstring keys")
+        self.config_key = f"x1plusd.modules.{self.driver_data.get('name', package)}"
+    
+    def load(self, *args, **kwargs):
+        # XXX: This currently does not load parent packages, and so
+        # therefore `from ..  import` and such will not work, and the
+        # package name is not a "real" package name in many ways.  Consider
+        # doing something that recurs upwards:
+        # https://docs.python.org/3/library/importlib.html#approximating-importlib-import-module
+        #
+        # Or consider registering a path entry hook to sandbox modules,
+        # maybe, in the future?  Then we could just use import, or
+        # something.
+        
+        assert not self.loaded
+        assert os.path.splitext(os.path.basename(self.path))[0] == self.package_name.split('.')[-1]
+
+        spec = importlib.util.spec_from_file_location(self.package_name, self.path)
+        module = importlib.util.module_from_spec(spec)
+        sys.modules[self.package_name] = module
+        spec.loader.exec_module(module)
+        
+        if hasattr(module, "load"):
+            module.load(*args, **kwargs)
+        self.module = module
+        self.loaded = True
+    
+    def start(self):
+        if hasattr(self.module, "start"):
+            self.module.start()
+
+def find_modules(directory: str, include_subdirs: bool = False, 
+                 base_package: str = "x1plus.services.x1plusd.modules",
+                 loader_type: str = "module"):
+    """Dynamically import all valid modules from a directory"""
+    modules = []
+
+    if not os.path.isdir(directory):
+        return modules
+    
+    pattern = f"{directory}/**[!__pycache__]**/[!_]*.py"
+
+    files = glob.glob(pattern, recursive=include_subdirs)
+    if include_subdirs:
+        pattern2 = f"{directory}/[!_]*.py"
+        files.extend(glob.glob(pattern2, recursive=False))
+
+    for file_path in files:
+        pkg_extension = file_path.replace(directory, "").replace("/", ".").replace(" ", "_").strip().lower()
+        if pkg_extension.startswith("."):
+            pkg_extension = pkg_extension[1:]
+        if pkg_extension.endswith(".py"):
+            pkg_extension = pkg_extension[:-3]
+
+        package = f"{base_package}.{pkg_extension}"
+
+        try:
+            modules.append(Module(file_path, package, loader_type = loader_type))
+        except ImportError:
+            # it just had no docstring keys; that's fine
+            pass
+        except Exception as e:
+            logger.error(f"failed to load module metadata from {file_path}: {e.__class__.__name__}: {e}")
+    return modules
+
+    
+def module_docstring_parser(filepath: str, loader_type: str) -> dict:
```

Comment:
back of my mind thought: rather than rolling our own parser, should we just yaml.safe_load?  that way we can have rich contents

Reply:
YAML implementation looks nice and is clean/easy to work with, I say it's in the right direction.